src/fsharp/TypedTreeOps.fs

```diff
@@ -2259,8 +2283,35 @@ let checkMemberVal membInfo arity m =
 let checkMemberValRef (vref: ValRef) =
     checkMemberVal vref.MemberInfo vref.ValReprInfo vref.Range
      
-let GetTopValTypeInCompiledForm g topValInfo ty m =
+/// Get information about the trait constraints for a set of typars.
+/// Put these in canonical order.
+let GetTraitConstraintInfosOfTypars g (tps: Typars) = 
+    [ for tp in tps do 
+            for cx in tp.Constraints do
+            match cx with 
+            | TyparConstraint.MayResolveMember(traitInfo, _) -> yield traitInfo 
+            | _ -> () ]
+    |> ListSet.setify (traitsAEquiv g TypeEquivEnv.Empty)
+    |> List.sortBy (fun traitInfo -> traitInfo.MemberName, traitInfo.ArgumentTypes.Length)
+
+/// Get information about the runtime witnesses needed for a set of generalized typars
+let GetTraitWitnessInfosOfTypars g numParentTypars tps = 
+    let tps = tps |> List.drop numParentTypars
+    let cxs = GetTraitConstraintInfosOfTypars g tps
+    cxs |> List.map (fun cx -> cx.TraitKey)
+
+/// Count the number of type parameters on the enclosing type
+let CountEnclosingTyparsOfActualParentOfVal (v: Val) = 
+    match v.ValReprInfo with 
+    | None -> 0
+    | Some _ -> 
+        if v.IsExtensionMember then 0
+        elif not v.IsMember then 0
+        else v.MemberApparentEntity.TyparsNoRange.Length
+
+let GetTopValTypeInCompiledForm g topValInfo numEnclosingTypars ty m =
     let tps, paramArgInfos, rty, retInfo = GetTopValTypeInFSharpForm g topValInfo ty m
+    let witnessInfos = GetTraitWitnessInfosOfTypars g numEnclosingTypars tps // TODO: parentTypars
```

Comment:
This has a "TODO", is if fair to expand on this a bit further? Will doing this "TODO" be for new language features in the future and/or break compat?

Reply:
I'll take a look

Reply:
I removed the TODO - it was fixed in an earlier iteration